dali/ft/filecopy.cpp

```diff
@@ -349,98 +349,115 @@ bool FileTransferThread::launchFtSlaveCmd()
+        for (;;)
+        {
+            msg.clear();
+            if (!catchReadBuffer(socket, msg, FTTIME_PROGRESS))
+                throwError1(RFSERR_TimeoutWaitSlave, url.str());
 
-        msg.setEndian(__BIG_ENDIAN);
-        msg.read(done);
-        if (done)
-            break;
+            msg.setEndian(__BIG_ENDIAN);
+            msg.read(done);
+            if (done)
+                break;
 
-        OutputProgress newProgress;
-        newProgress.deserializeCore(msg);
-        newProgress.deserializeExtra(msg, 1);
-        newProgress.deserializeExtra(msg, 2);
-        sprayer.updateProgress(newProgress);
+            OutputProgress newProgress;
+            newProgress.deserializeCore(msg);
+            newProgress.deserializeExtra(msg, 1);
+            newProgress.deserializeExtra(msg, 2);
+            sprayer.updateProgress(newProgress);
 
-        LOG(MCdebugProgress(10000), job, "Update %s: %d %" I64F "d->%" I64F "d", url.str(), newProgress.whichPartition, newProgress.inputLength, newProgress.outputLength);
-        if (isAborting())
-        {
-            if (!sendRemoteAbort(socket))
-                throwError1(RFSERR_TimeoutWaitSlave, url.str());
+            LOG(MCdebugProgress(10000), job, "Update %s: %d %" I64F "d->%" I64F "d", url.str(), newProgress.whichPartition, newProgress.inputLength, newProgress.outputLength);
+            if (isAborting())
+            {
+                if (!sendRemoteAbort(socket))
+                    throwError1(RFSERR_TimeoutWaitSlave, url.str());
+            }
         }
-    }
-    msg.read(ok);
-    setErrorOwn(deserializeException(msg));
+        msg.read(ok);
+        setErrorOwn(deserializeException(msg));
 
-    LOG(MCdebugProgressDetail, job, "Finished generating part %s [%p] ok(%d) error(%d)", url.str(), this, (int)ok, (int)(error!=NULL));
+        LOG(MCdebugProgressDetail, job, "Finished generating part %s [%p] ok(%d) error(%d)", url.str(), this, (int)ok, (int)(error!=NULL));
 
-    // if communicating with ftslave, the process has a final ack wait
-    if (sprayer.useFtSlave)
+        // if communicating with ftslave, the process has a final ack wait
+        if (sprayer.useFtSlave)
+        {
+            // ftslave sends back a final ack.
+            msg.clear().append(true);
+            catchWriteBuffer(socket, msg);
+            if (sprayer.options->getPropInt("@fail", 0)) // probably used as a debug testing option
+                throwError1(DFTERR_CopyFailed, "unknown reason");
+        }
+    }
+    catch (IException *e)
     {
-        // ftslave sends back a final ack.
-        msg.clear().append(true);
-        catchWriteBuffer(socket, msg);
-        if (sprayer.options->getPropInt("@fail", 0)) // probably used as a debug testing option
-            throwError1(DFTERR_CopyFailed, "unknown reason");
+        exception.set(e);
+    }
+    if (localFtSlaveHandle)
+    {
+        DWORD runcode;
+        if (!wait_program_timeout(localFtSlaveHandle, runcode, 5000))
```

Comment:
>This called after we know the child has done all its work and should have ended very recently tho, correct ?

Yes.

>Calling wait_program_timeout() to prevent a zombie is great and needed, but 5 secs seems arbitrary ?

It is a bit arbitrary, less time should be good, 5s was just some headroom. Not sure what most sensible value is here.